src/main/scala/beam/agentsim/agents/ridehail/ParallelAlonsoMoraAlgForRideHail.scala

```diff
@@ -0,0 +1,135 @@
+  demand: List[CustomerRequest],
+  supply: List[VehicleAndSchedule],
+  timeWindow: Map[MobilityServiceRequestType, Int],
+  radius: Int,
+  implicit val skimmer: BeamSkimmer
+) {
+
+  private def parallelBuildOfRTVGraph(): AlonsoMoraPoolingAlgForRideHail.RTVGraph = {
+    import scala.concurrent.ExecutionContext.Implicits.global
+    import scala.collection.mutable.{ListBuffer => MListBuffer}
+    val rTvG = AlonsoMoraPoolingAlgForRideHail.RTVGraph(classOf[DefaultEdge])
+    Future.sequence(supply.withFilter(_.getFreeSeats > 0).map { v =>
+      Future {
+        val vertices = MListBuffer.empty[RTVGraphNode]
+        val edges = MListBuffer.empty[(RTVGraphNode, RTVGraphNode)]
+        val finalRequestsList = MListBuffer.empty[RideHailTrip]
+        demand
+          .withFilter(
+            x =>
+              AlonsoMoraPoolingAlgForRideHail.getTimeDistanceAndCost(v.schedule.head, x.pickup).distance.get <= radius
+          )
+          .foreach(
+            r =>
+              AlonsoMoraPoolingAlgForRideHail
+                .getRidehailSchedule(timeWindow, v.schedule ++ List(r.pickup, r.dropoff)) match {
+                case Some(schedule) =>
+                  val t = RideHailTrip(List(r), schedule)
+                  finalRequestsList append t
+                  if (!vertices.contains(v)) vertices.append(v)
+                  vertices.append(r, t)
+                  edges.append((r, t), (t, v))
+                case _ =>
+            }
+          )
+        if (finalRequestsList.nonEmpty) {
+          for (k <- 2 until v.getFreeSeats + 1) {
+            var index = 1
+            val kRequestsList = MListBuffer.empty[RideHailTrip]
+            for (t1 <- finalRequestsList) {
+              for (t2 <- finalRequestsList
+                     .drop(index)
+                     .withFilter(
+                       x =>
+                         !(x.requests exists (s => t1.requests contains s)) && (t1.requests.size + x.requests.size) == k
+                     )) {
+                AlonsoMoraPoolingAlgForRideHail
+                  .getRidehailSchedule(
+                    timeWindow,
+                    v.schedule ++ (t1.requests ++ t2.requests).flatMap(x => List(x.pickup, x.dropoff))
+                  ) match {
+                  case Some(schedule) =>
+                    val t = RideHailTrip(t1.requests ++ t2.requests, schedule)
+                    kRequestsList.append(t)
+                    vertices.append(t)
+                    t.requests.foldLeft(()) { case (_, r) => edges.append((r, t)) }
+                    edges.append((t, v))
+                  case _ =>
+                }
+              }
+              index += 1
+            }
+            finalRequestsList.appendAll(kRequestsList)
+          }
+        }
+        (vertices, edges)
+      }
+    }) onComplete {
+      case Success(result) =>
+        result foreach {
+          case (vertices, edges) =>
+            vertices foreach (vertex => rTvG.addVertex(vertex))
```

Comment:
First, good job not letting mutability enter the Future itself. However this completion may not work the way you think because you are then mutating data afterwards*. Yes, it's on the return thread, however onComplete is not blocking, so when you return `rTvG` you are returning it immediately. Later the onComplete will kick in and in theory may update the edge for the reference you passed out, but it gets shady. I see two options - block until the future is truly complete (in which case you should use map/foreach instead of onComplete to make sure ALL work is done) or return the Future itself, which will return the rTvG. Then the code that needs the final value can be the one to wait - maybe it doesn't need it until way later anyway?

*Again good that it is all bunched onto the same thread via the packaging of Future.sequence, though.

Reply:
I just pushed a fix for this - go ahead and take a look. The functionality shouldn't have changed, just the level where variables were manipulated in the future, and that the end result is a Future, which is just being awaited in the test.